### ISSUE TITLE
AssociatorNames Result class parameter Error return from OpenPegasus

### ISSUE THREAD

**Issue by KSchopmeyer** [pywbem/pywbem#1907]
OpenPegasus returns an error response from the following AssociatorNames request that includes a ResultClass name that is for a class in a different namespace than the request source.
The definition of the AssociatorNames ResultClass parameter is that it can be either a String or CIMClassName.

Note that the application passes an instance of CIMCLassName to the AssociatorNames call. The resullting HTML sends the classname component to the server.

The server responds:

```
    <IMETHODRESPONSE NAME="AssociatorNames">
    <ERROR CODE="5" DESCRIPTION="CIM_ERR_INVALID_CLASS: Test_StorageSystem"/>
```

I am assuming that this is an OpenPegasus error in that DSP0200 defines that :

"The ResultClass input parameter, if not NULL, shall be a valid CIM class name. It acts as a filter on the returned set of names by mandating that each returned name identify an object that shall be either an instance of this class (or one of its subclasses) or be this class (or one of its subclasses)."

I.e. is is a class name, not a classname with namespace.   

Within the Pegasus server, pywbemcli finds:

```
    pywbemcli> class find Test_MCSStorageSystem                                     
      test/TestProvider:Test_MCSStorageSystem

    pywbemcli> instance count Test_MCSStorageSystem                                 
    Count of instances per class
    +-------------------+-----------------------+---------+
    | Namespace         | Class                 |   count |
    |-------------------+-----------------------+---------|
    | test/TestProvider | Test_MCSStorageSystem |       6 |
    +-------------------+-----------------------+---------+
```

so the class and instances do exist but not in the source namespace.  I.e. this defines an association that crosses namespaces.

The question is whether this is an error in OpenPegasus where the server should be not taking namespace into account when it is running the ResultClass filter on this particular AssociationName request or whether the namespace should be part of the request to the server.

```
  2019-10-22 11:59:35,685-pywbem.api.1-4793-Request:1-4793 AssociatorNames(
  AssocClass=CIMClassName(classname='PG_ElementConformsToProfile',   namespace='root/PG_Interop', host='leonard'),
  ObjectName=CIMInstanceName(classname='PG_RegisteredProfile', keybindings=NocaseDict({'InstanceID': 'SNIA+Array+1.1.0'}), namespace='root/PG_Interop', host='leonard'),
  ResultClass=CIMClassName(classname='Test_StorageSystem', namespace='test/TestProvider', host='leonard'),
  ResultRole='ManagedElement', Role='ConformantStandard')

  2019-10-22 11:59:35,687-pywbem.http.1-4793-Request:1-4793 POST /cimom 11 http://localhost CIMOperation:'MethodCall' CIMMethod:'AssociatorNames' CIMObject:'root/PG_Interop'
    <?xml version="1.0" encoding="utf-8" ?>
  <CIM CIMVERSION="2.0" DTDVERSION="2.0"><MESSAGE ID="1001" PROTOCOLVERSION="1.0">
  <SIMPLEREQ>
  <IMETHODCALL NAME="AssociatorNames"><LOCALNAMESPACEPATH>
    <NAMESPACE NAME="root"/><NAMESPACE NAME="PG_Interop"/>
    </LOCALNAMESPACEPATH>
   <IPARAMVALUE NAME="ObjectName">
   <INSTANCENAME CLASSNAME="PG_RegisteredProfile">  <KEYBINDING NAME="InstanceID">
  <KEYVALUE  VALUETYPE="string">SNIA+Array+1.1.0</KEYVALUE></KEYBINDING>
  </INSTANCENAME></IPARAMVALUE>
  <IPARAMVALUE NAME="AssocClass"><CLASSNAME  NAME="PG_ElementConformsToProfile"/>
  </IPARAMVALUE>
  <IPARAMVALUE NAME="ResultClass"><CLASSNAME NAME="Test_StorageSystem"/> 
  </IPARAMVALUE>
  <IPARAMVALUE NAME="Role"><VALUE>ConformantStandard</VALUE></IPARAMVALUE> 
  <IPARAMVALUE NAME="ResultRole"><VALUE>ManagedElement</VALUE></IPARAMVALUE>
  </IMETHODCALL></SIMPLEREQ></MESSAGE></CIM>

2019-10-22 11:59:35,689-pywbem.http.1-4793-Response:1-4793 200:OK 11 Content-Type:'application/xml; charset=utf-8' content-length:'0000000298' CIMOperation:'MethodResponse'
    <?xml version="1.0" encoding="utf-8" ?>
 <CIM CIMVERSION="2.0" DTDVERSION="2.0">
   <MESSAGE ID="1001" PROTOCOLVERSION="1.0">
  <SIMPLERSP>
  <IMETHODRESPONSE NAME="AssociatorNames">
  <ERROR CODE="5" DESCRIPTION="CIM_ERR_INVALID_CLASS: Test_StorageSystem"/>  
  </IMETHODRESPONSE>
  </SIMPLERSP>
  </MESSAGE>
  </CIM>
```

**Comment by KSchopmeyer** [comment]
Note that OpenPegasuc CLI returns the same error:

kschopmeyer@leonard:~/pgogtest/pegasus$ cimcli an PG_RegisteredProfile.InstanceID=\"SNIA+Array+1.1.0\" -n root/PG_Interop
10 associatorNames
//leonard/root/PG_InterOp:PG_RegisteredSubProfile.InstanceId="SNIA+Block Server Performance+1.1.0"
//leonard/root/PG_InterOp:PG_RegisteredSubProfile.InstanceId="SNIA+Disk Drive Lite+1.1.0"
//leonard/root/PG_InterOp:PG_RegisteredSubProfile.InstanceId="SNIA+Indication+1.1.0"
//leonard/root/PG_InterOp:PG_SoftwareIdentity.InstanceId="TestServerProfileProvidersModule+StorageSystemTestProvider"
//leonard/root/PG_InterOp:PG_SoftwareIdentity.InstanceId="TestServerProfileProvidersModule+MCSStorageSystemTestProvider"
//leonard/root/PG_InterOp:PG_RegisteredProfile.InstanceId="SNIA+SMI-S+1.2.0"
//leonard/test/TestProvider:Test_StorageSystem.CreationClassName="Test_StorageSystem",Name="StorageSystemInstance1"
//leonard/test/TestProvider:Test_StorageSystem.CreationClassName="Test_StorageSystem",Name="StorageSystemInstance2"
//leonard/test/TestProvider:Test_MCSStorageSystem.CreationClassName="Test_MCSStorageSystem",Name="MCSStorageSystemInstance1"
//leonard/test/TestProvider:Test_MCSStorageSystem.CreationClassName="Test_MCSStorageSystem",Name="MCSStorageSystemInstance2"
kschopmeyer@leonard:~/pgogtest/pegasus$ cimcli an PG_RegisteredProfile.InstanceID=\"SNIA+Array+1.1.0\" -n root/PG_Interop -rc Test_MCSStorageSystem
cimcli CIMException:  Cmd= an Object= PG_RegisteredProfile.InstanceID="SNIA+Array+1.1.0" Code= 5
CIM_ERR_INVALID_CLASS: Test_MCSStorageSystem
kschopmeyer@leonard:~/pgogtest/pegasus$ cimcli an PG_RegisteredProfile.InstanceID=\"SNIA+Array+1.1.0\" -n root/PG_Interop -rc Test_MCSStorageSystem
cimcli CIMException:  Cmd= an Object= PG_RegisteredProfile.InstanceID="SNIA+Array+1.1.0" Code= 5
CIM_ERR_INVALID_CLASS: Test_MCSStorageSystem

**Comment by KSchopmeyer** [comment]
In retrospect, the question may come down to whether there is a requirement for the class for all references to exist in both namespaces if an association crosses namespaces.  That may be a pegasus question as much as a spec question.  

This comes from the requirement for the associators with class method rather than the associator instance method.

**Comment by andy-maier** [comment]
For associations that cross namespaces, the association class needs to exist in both namespaces. The source end class only needs to exist in the source namespace, and the far end class only needs to exist in the target namespace.

I'm not sure anymore whether we just discussed it that way, or also wrote it in some spec. Did not have time to check it out. But it only works that way.

So one issue in your server setup is that the association class does not exist in both namespaces.

What we do about the fact that pywbem accepts a CIMClassName object for the ResultClass but then ignores host and namespace, is to be discussed. It could be a host and namespace that is neither the source nor the target namespace. The options for pywbem are:
* Just describe that host and namespace are being ignored.
* Check that host and namespace if specified in the CIMClassName object are either for source namespace/host or for target namespace/host, and raise some user exception otherwise.

**Comment by andy-maier** [comment]
Discussed this in the call today, and we realized that the target namespace is not known before the call is made, so it cannot be checked against upfront.

**Comment by andy-maier** [comment]
The clarification in the docs that host/namespace are ignored, was addressed with PR #1922 for master, and TBD for 0.15.0.
